src/cuckoocache.h

```diff
@@ -0,0 +1,450 @@
+     * true.
+     */
+    bit_packed_atomic_flags(uint32_t size)
+    {
+        // pad out the size if needed
+        size = (size + 7) / 8;
+        mem.reset(new std::atomic<uint8_t>[size]);
+        for (uint32_t i = 0; i < size; ++i)
+            mem[i].store(0xFF);
+    };
+
+    /** setup marks all entries and ensures that bit_packed_atomic_flags can store
+     * at least size entries
+     *
+     * @param b the number of elements to allocate space for
+     * @post bit_set, bit_unset, and bit_is_set function properly forall x. x <
+     * b
+     * @post All calls to bit_is_set (without subsequent bit_unset) will return
+     * true.
+     */
+    inline void setup(uint32_t b)
+    {
+        bit_packed_atomic_flags d(b);
+        std::swap(mem, d.mem);
+    }
+
+    /** bit_set sets an entry as discardable. 
+     *
+     * @param s the index of the entry to bit_set.
+     * @post immediately subsequent call (assuming proper external memory
+     * ordering) to bit_is_set(s) == true.
+     *
+     */
+    inline void bit_set(uint32_t s)
+    {
+        mem[s >> 3].fetch_or(1 << (s & 7), std::memory_order_relaxed);
+    }
+
+    /**  bit_unset marks an entry as something that should not be overwritten  
+     *
+     * @param s the index of the entry to bit_unset.
+     * @post immediately subsequent call (assuming proper external memory
+     * ordering) to bit_is_set(s) == false.
+     */
+    inline void bit_unset(uint32_t s)
+    {
+        mem[s >> 3].fetch_and(~(1 << (s & 7)), std::memory_order_relaxed);
+    }
+
+    /** bit_is_set queries the table for discardability at s
+     *
+     * @param s the index of the entry to read.
+     * @returns if the bit at index s was set.
+     * */
+    inline bool bit_is_set(uint32_t s) const
+    {
+        return (1 << (s & 7)) & mem[s >> 3].load(std::memory_order_relaxed);
+    }
+};
+
+/** cache implements a cache with properties similar to a cuckoo-set
+ *
+ *  The cache is able to hold up to (~(uint32_t) 1) elements.
+ *
+ *  Read Operations:
+ *      - contains(*, false)
+ *
+ *  Read/Erase Operations:
+ *      - contains(*, true)
+ *      - allow_erase()
+ *
+ *  Write Operations:
+ *      - setup()
+ *      - setup_bytes()
+ *      - insert()
+ *      - please_keep()
+ *      - rehash()
+ *
+ *  Synchronization Free Operations:
+ *      - invalid()
+ *      - compute_hashes()
+ *
+ * User Must Guarantee:
+ *
+ * 1) Write Requires synchronized access (e.g., a lock)
+ * 2) Read Requires no concurrent Write, synchronized with the last insert.
+ * 3) Erase requires no concurrent Write, synchronized with last insert.
+ * 4) An Eraser must release all Erases before allowing a new Writer.
+ *
+ *
+ * Note on function names:
+ *   - The name "allow_erase" is used because the real discard happens later.
+ *   - The name "please_keep" is used because elements may be erased anyways on insert.
+ *
+ * @tparam Element should be a POD type that is 32-alignable
```

Comment:
I suppose the way it is currently written (using swap and move) it should be safe to remove both requirements (POD and 32-alignable). 

Prior versions of the code I think were unsafe for that use case.